### PR TITLE
Fix event functions chain crash

### PR DESCRIPTION
**Pull request by sebjulliand**
### Changes
Resolves https://github.com/codefori/vscode-ibmi/issues/2181
This PR prevents a failing event function from crashing the whole chain (i.e. preventing other functions from being run when connecting/disconnecting).

### How to test this PR
1. Connect
2. Make sure every view loads correctly

Connecting to PUB400 should show a warning about the failing function.

### Checklist
* [x] have tested my change

### CODE REVIEW

File: src/api/Instance.ts [hunk 1]
```diff
@@ -66,7 +66,13 @@ export default class Instance {
 
   async processEvent(event: IBMiEvent) {
     for (const subscriber of this.subscribers.filter(s => s.event === event)) {
-      await subscriber.func();
+      try{
+        await subscriber.func();
+      }
+      catch(error){
+        const name = subscriber.func.name ? `Event function ${subscriber.func.name}` : `Anonymous event function`;
+        vscode.window.showWarningMessage(`${name} failed: ${error}`)
```
Comment (worksofliam) [inline src/api/Instance.ts:74]: I don't know if the user cares. I think we should `console.error` here and let the using-extension handle the error in their function (e.g. show any warnings/issues to the user.)

**Review by worksofliam — CHANGES_REQUESTED** [review]

File: src/api/Instance.ts (hunk 1 above)
Comment (sebjulliand) [reply src/api/Instance.ts:74]: Fair enough!

**Comment by sebjulliand** [comment]
@worksofliam there you go.
I added a `subscribe` function that supersedes `onEvent` and replaced every occurrence of it.
I also added some time tracing around these function calls...it won't hurt, in case of a bottleneck situation.
And I also removed a few redundant view refreshes that were called twice.

**Comment by worksofliam** [comment]
@sebjulliand Initial review of the code looks good. Glad to see `onEvent` marked as depreciated and replaced with `subscribe`. I will have a go at testing it tomorrow. I believe after this has merged, and with the release of 3.0.0, we will need to update the [documentation on our API](https://github.com/codefori/vscode-ibmi/pull/2182/files). After merging, and release of the next version, I can go ahead and try out `subscribe` in the Db2 for i extension.

File: src/api/Instance.ts [hunk 2]
```diff
@@ -56,17 +59,57 @@ export default class Instance {
...
+    }
+    return eventSubscribers;
+  }
+
+  /**
+   * @deprecated Will be removed in `v3.0.0`; use {@link subscribe} instead
+   */
   onEvent(event: IBMiEvent, func: Function): void {
-    this.subscribers.push({ event, func });
+    this.getSubscribers(event).set(`deprecated - ${func.name || "unknown"}_${this.deprecationCount++}`, { func });
```
Comment (worksofliam) [inline src/api/Instance.ts:88]: I think an additional `console.log` in here so extension developers can see `onEvent` is going away in 3.0.0.

Something like `you are using onEvent, which is deprecated and will be removed in 3.0.0. Please use `subscribe`

**Review by worksofliam — APPROVED** [review]
I think one last change and we should be good to go here.

File: src/api/Instance.ts (hunk 2 above)
Comment (sebjulliand) [reply src/api/Instance.ts:88]: Good idea!

**Comment by sebjulliand** [comment]
Tgabjs for the review @worksofliam . Give it a last whirl and I guess we'll be good to go!

**Comment by worksofliam** [comment]
@sebjulliand I can confirm this PR fixed a startup bug with the database extension as the event error handler is working as expected.

image

**Review by worksofliam — APPROVED** [review]

**Comment by worksofliam** [comment]
@sebjulliand Merge when you are ready! Thanks for your work on this.

**Pull request merged**